salt/_modules/metalk8s.py

```diff
@@ -355,3 +358,164 @@ def cmp_sorted(*args, **kwargs):
         kwargs['key'] = functools.cmp_to_key(kwargs.pop('cmp'))
 
     return sorted(*args, **kwargs)
+
+
+def _error(ret, err_msg):
+    ret["result"] = False
+    ret["comment"] = err_msg
+    return ret
+
+
+def _atomic_copy(source, dest, tmp_prefix="."):  # pragma: no cover
+    """Minimalistic implementation of an atomic copy operation.
+
+    First, we write the contents of the source file to a temporary file, which
+    is saved in the same directory as the target, with a custom prefix.
+    Finally, we link the temporary file contents to the destination filename.
+    """
+    base_name = os.path.basename(dest)
+    dir_name = os.path.dirname(dest)
+
+    with open(source, mode='rb') as src_file:
+        contents = src_file.read()
+
+    with tempfile.NamedTemporaryFile(
+        prefix=tmp_prefix, dir=dir_name, delete=False,
+    ) as tmp_file:
+        tmp_file.write(contents)
```

Comment:
You'll want to `fsync` `tmp_file`.

Reply:
It's flushed on `.close()`, which the contextmanager takes care of, IIUC.

See:
- https://github.com/python/cpython/blob/master/Lib/tempfile.py#L427-L434
- https://github.com/python/cpython/blob/master/Lib/tempfile.py#L486-L491

But I can add a `tmp_file.flush()` if you think that's required.

Reply:
A `close` (and `flush`, for buffered IO) doesn't `fsync` data to disk.

Reply:
OK, adding `os.fdatasync(tmp_file.fileno())` in that case.

Reply:
You want `fsync`, not `fdatasync`, since with `fdatasync` the perms you set before are not guaranteed to be synced to disk.

Reply:
Ha, my bad I didn't know the difference, thanks!